src/Microsoft.AspNet.Html.Abstractions/HtmlContentBuilderExtensions.cs

```diff
@@ -132,5 +180,145 @@ private string DebuggerToString()
+        {
+            private readonly IFormatProvider _formatProvider;
+            private readonly string _format;
+            private readonly object[] _args;
+
+            public HtmlFormatString(string format, object[] args)
+                : this(null, format, args)
+            {
+            }
+
+            public HtmlFormatString(IFormatProvider formatProvider, string format, object[] args)
+            {
+                if (format == null)
+                {
+                    throw new ArgumentNullException(nameof(format));
+                }
+
+                _formatProvider = formatProvider ?? CultureInfo.CurrentCulture;
+                _format = format;
+                _args = args;
+            }
+
+            public void WriteTo(TextWriter writer, IHtmlEncoder encoder)
+            {
+                if (writer == null)
+                {
+                    throw new ArgumentNullException(nameof(writer));
+                }
+
+                if (encoder == null)
+                {
+                    throw new ArgumentNullException(nameof(encoder));
+                }
+
+                var formatProvider = new EncodingFormatProvider(_formatProvider, encoder);
+                var builder = new StringBuilder();
+                builder.AppendFormat(formatProvider, _format, _args);
+
+                for (var i = 0; i < builder.Length; i++)
+                {
+                    writer.Write(builder[i]);
+                }
+            }
+
+            private string DebuggerToString()
+            {
+                using (var writer = new StringWriter())
+                {
+                    WriteTo(writer, HtmlEncoder.Default);
+                    return writer.ToString();
+                }
+            }
+        }
+
+        // This class implements Html encoding via an ICustomFormatter. Passing an instance of this
+        // class into a string.Format method or anything similar will evaluate arguments implementing
+        // IHtmlContent without HTML encoding them, and will give other arguments the standard
+        // composite format string treatment, and then HTML encode the result.
+        //
+        // Plenty of examples of ICustomFormatter and the interactions with string.Format here:
+        // https://msdn.microsoft.com/en-us/library/system.string.format(v=vs.110).aspx#Format6_Example
+        private class EncodingFormatProvider : IFormatProvider, ICustomFormatter
+        {
+            private IHtmlEncoder _encoder;
+            private IFormatProvider _formatProvider;
+
+            public EncodingFormatProvider(IFormatProvider formatProvider, IHtmlEncoder encoder)
+            {
+                _formatProvider = formatProvider;
+                _encoder = encoder;
+            }
+
+            public string Format(string format, object arg, IFormatProvider formatProvider)
+            {
+                // This is the case we need to special case, and avoid encoding.
```

Comment:
How does this bit avoid encoding? We're just calling through to `.WriteTo` on an `IHtmlContent` which may or may not encode.